core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1508,4 +1519,37 @@ class ReplicaManager(val config: KafkaConfig,
       tp -> epochEndOffset
     }
   }
+
+  def electPreferredLeaders(controller: KafkaController,
+                            partitions: Set[TopicPartition],
+                            responseCallback: Map[TopicPartition, ApiError] => Unit,
+                            requestTimeout: Long): Unit = {
+
+    val (validPartitions, invalidPartitions) = partitions.partition(tp => metadataCache.contains(tp))
+
+    val invalidPartitionsResults = invalidPartitions.map { p =>
+      val msg = s"Skipping preferred replica leader election for partition ${p} since it doesn't exist."
+      logger.info(msg)
+      p -> new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION, s"The partition does not exist.")
+    }.toMap
+
+    def electionCallback(waiting: Set[TopicPartition], results: Map[TopicPartition, ApiError]) = {
+      if (waiting.nonEmpty) {
+        // timeout
+        val expectedLeaders = waiting.map(
+          tp => ElectPreferredLeaderMetadata(tp, controller.controllerContext.partitionReplicaAssignment(tp).head))
```

Comment:
Yes, I meant the latter. Controller.controllerContext is supposed to be internal to the controller. It would be useful not to use it outside of the controller.

Reply:
As discussed earlier, it would be useful not to use Controller.controllerContext here since it's meant to be internal to the controller.